### ISSUE TITLE
window.render() terminates the app

### ISSUE THREAD

**Issue by QwertyR0** [kmamal/node-sdl#4]
Hello, i am trying to make a simple game engine but when i run window.render in my game loop it terminates the app
here is the full code:
``` ts
async function gameLoop(this: any){
	
		var dt: number = 0;
		var now: number = 0;
		while(true){
			now = Date.now();
			dt += now - this.lastUpdate;
			//console.log(`${now}, ${this.lastUpdate}`); for debugging
			this.lastUpdate = now;
			while (dt >= 1) {
				this.levels[0].update(this.window, dt);
				dt--;
			}  
			if(!this.window.destroyed){
				await this.levels[0].render();
			}
		
			const buffer = this.canvas.toBuffer('raw');
			this.window.render(this.width, this.height, this.width * 4, 'rgba32', buffer);
		}
}
```

**Comment by kmamal** [comment]
@QwertyR0 I can't really tell what the problem is from only that part of your code. Are there ary error messages in the terminal when the app terminates? Do you have the full code in a repo somewhere so I can check?

**Comment by QwertyR0** [comment]
> @QwertyR0 I can't really tell what the problem is from only that part of your code. Are there ary error messages in the terminal when the app terminates? Do you have the full code in a repo somewhere so I can check?

no it just closes

**Comment by QwertyR0** [comment]
> @QwertyR0 I can't really tell what the problem is from only that part of your code. Are there ary error messages in the terminal when the app terminates? Do you have the full code in a repo somewhere so I can check?

the full code:
```js
const sdl = require("@kmamal/sdl");
const Canvas = require('canvas');
const graphics = require("./render/Graphics");

async function render(){
	if(!this.window.destroyed){
		this.ctx.clearRect(0, 0, this.width, this.height);

		this.levels[0].render(graphics)

		const buffer = this.canvas.toBuffer('raw');
		this.window.render(this.width, this.height, this.width*4, 'bgra32', buffer);
	}
}

async function gameLoop(){
	var dt = 0;
	var now = 0;
	while(this.running){
		await setTimeout(()=>{}, 100);
		now = Date.now();
		dt = now - this.lastUpdate;
		this.lastUpdate = now;

		while (dt >= 1) {
			this.levels[0].update(this.window, dt);
			dt--;
		}

		render.bind(this)();
	}
}

module.exports = class {
	levels = [];
	constructor(width, height, title){
		this.width = width;
		this.lastUpdate = 0;
		this.height = height;
		this.running = false;
		this.title = title;
		this.window = sdl.video.createWindow({ title: this.title });
		this.canvas = Canvas.createCanvas(this.width, this.height);
		this.ctx = this.canvas.getContext('2d');
	}

	addLevel(level) {
		this.levels.push(level);
	}

	loadLevel(levelID) {
		this.levels.forEach((value, index)=> {
			if(value.id === levelID){
				this.levels.splice(0, 0, this.levels.splice(index, 1)[0]);
			}
		});
	}

	run() {
		this.running = true;
		this.levels[0].init(this.window);
		try{
			this.lastUpdate = Date.now();
			gameLoop.bind(this)();
		} catch(e) {
			console.error(e);
		}
	}

	pause() {
		this.running = false;
	}
}
```

**Comment by kmamal** [comment]
@QwertyR0 I'm assuming you are building this project as a way to better learn JavaScript programming? I would like to point out a few issues with your code:

1. There is no reason for the `render` function to be declared `async`. You can remove the `async` keyword from `async function render(){...}`.
2. The default `setTimeout()` function does not return a promise. It returns the id for the timeout that you started. So `await setTimeout(() => {}, 100)` will not do what you think it should do. What you really want to do is to import the promise-based `setTimeout` by adding `const { setTimeout } = require('timers/promises')` to your imports and then use it like `await setTimeout(100)`.
3. You are calling your "private" functions like `func.bind(this)()`. This works, but the `.bind()` call creates a new function each time it is called. It is better to call these functions like `func.call(this)` and if you want to pass arguments like `func.call(this, arg1, arg2)`.
4. In your `loadLevel` function I believe you are trying to move the level with `value.id === levelID` to the beginning? The code you have will (probably) work but it looks very weird. A better alternative would be:
```js
const index = this.levels.findIndex((level) => level.id === levelID)
this.levels.unshift(this.levels.splice(index, 1)[0])
```

Other than that, this part of your program looks fine and I don't see how `window.render()` is causing the app to terminate. Here's some things you can do to help me debug this:
1. Paste here any error messages your program generates. When running your program from the command line (such as with `node index.js`) you should get an error message printed if your program exited unsuccessfully.
2. Add some debug statements like `console.log(this.window, this.width, this.height, this.width*4, 'bgra32', buffer)` before the function call and let me know what it prints.
3. I have a suspicion that your problem might be in one of the `level` functions(`init`, `update`, or `render`). You could try wrapping these functions in `try/catch` statements to make sure they are not causing problems.

**Comment by QwertyR0** [comment]
> @QwertyR0 I'm assuming you are building this project as a way to better learn JavaScript programming? If you don't mind, I'd like to point out a few issues with your code:
> 
> 1. There is no reason for the `render` function to be declared `async`. You can remove the `async` keyword from `async function render(){...}`.
> 2. The default `setTimeout()` function does not return a promise. It returns the id for the timeout that you started. So `await setTimeout(() => {}, 100)` will not do what you think it should do. What you really want to do is to import the promise-based `setTimeout` by adding `const { setTimeout } = require('timers/promises')` to your imports and then use it like `await setTimeout(100)`.
> 3. You are calling your "private" functions like `func.bind(this)()`. This works, but the `.bind()` call creates a new function each time it is called. It is better to call these functions like `func.call(this)` and if you want to pass arguments like `func.call(this, arg1, arg2)`.
> 4. In your `loadLevel` function I believe you are trying to move the level with `value.id === levelID` to the beginning? The code you have will (probably) work but it looks very weird. A better alternative would be:
> 
> ```js
> const index = this.levels.findIndex((level) => level.id === levelID)
> this.levels.unshift(this.levels.splice(index, 1)[0])
> ```
> 
> Other than that, this part of your program looks fine and I don't see how `window.render()` is causing the app to terminate. Here's some things you can do to help me debug this:
> 
> 1. Paste here any error messages your program generates. When running your program from the command line (such as with `node index.js`) you should get an error message printed if your program exited unsuccessfully.
> 2. Add some debug statements like `console.log(this.window, this.width, this.height, this.width*4, 'bgra32', buffer)` before the function call and let me know what it prints.
> 3. I have a suspicion that your problem might be in one of the `level` functions(`init`, `update`, or `render`). You could try wrapping these functions in `try/catch` statements to make sure they are not causing problems.

Thanks! i will try those. Actually the code was much cleaner and better but i ran out of ideas while trying fix this issue and tried stupid things. Again thanks for helping!

**Comment by QwertyR0** [comment]
Hello again! I tried the things that you told me.
1. As i said there are no errors popping in the command line.
2. This is the output that it gave:
`
Window {
  _events: [Object: null prototype] {
    removeListener: [Function (anonymous)],
    newListener: [Function (anonymous)],
    close: [Function: bound onceWrapper] { listener: [Function (anonymous)] }
  },
  _eventsCount: 3,
  _maxListeners: undefined,
  _validEvents: [
    'show',          'hide',
    'expose',        'minimize',
    'maximize',      'restore',
    'move',          'resize',
    'focus',         'blur',
    'hover',         'leave',
    'beforeClose',   'close',
    'keyDown',       'keyUp',
    'textInput',     'mouseButtonDown',
    'mouseButtonUp', 'mouseMove',
    'mouseWheel',    'dropBegin',
    'dropText',      'dropFile',
    'dropComplete'
  ],
  _id: 1,
  _x: 125,
  _y: 389,
  _width: 640,
  _height: 480,
  _native: <Buffer e8 06 35 00 00 00 00 00>,
  _title: 'hello',
  _visible: true,
  _fullscreen: false,
  _resizable: false,
  _borderless: false,
  _opengl: false,
  _focused: true,
  _hovered: false,
  _minimized: false,
  _maximized: false,
  _destroyed: false,
  [Symbol(kCapture)]: false
} 128 128 512 bgra32 <Buffer 00 00 00 40 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 00 00 ... 65486 more bytes>
`

3. It seems like there are no problems with them.

thanks!

**Comment by kmamal** [comment]
I think that looks ok so we'll have to dig a bit deeper.

I tried running your code by naming the class `Test` and testing like this:
```
const g = new Test(640, 480, "TEST")
g.addLevel({
	init: () => {},
	update: () => {},
	render: () => {},
})
g.run()
```

This runs fine. A window is created and stays up.

May I ask you to share more of your code? Can you maybe create a github repository an upload the whole thing?

**Comment by QwertyR0** [comment]
> I think that looks ok so we'll have to dig a bit deeper.
> 
> I tried running your code by naming the class `Test` and testing like this:
> 
> ```js
> const g = new Test(640, 480, "TEST")
> g.addLevel({
>   init: () => {},
>   update: () => {},
>   render: () => {},
> })
> g.run()
> ```
> 
> This runs fine. A window is created and stays up.
> 
> May I ask you to share more of your code? Can you maybe create a github repository an upload the whole thing?

here it is:
https://github.com/QwertyR0/Qngine

**Comment by kmamal** [comment]
Aaaaah I see now! The dimensions of the canvas must match the dimensions of the window you render it on. You need to pass width and height when creating the window like this:
```js
sdl.video.createWindow({
  width: this.width,
  height: this.height,
  title: this.title,
})
```

In the examples I get the canvas width and height directly from the corresponding window properties so it works fine. I should update the docs about this or maybe try to make it behave nicer in future versions.

**Comment by QwertyR0** [comment]
> Aaaaah I see now! The dimensions of the canvas must match the dimensions of the window you render it on. You need to pass width and height when creating the window like this:
> 
> ```js
> sdl.video.createWindow({
>   width: this.width,
>   height: this.height,
>   title: this.title,
> })
> ```
> 
> In the examples I get the canvas width and height directly from the corresponding window properties so it works fine. I should update the docs about this or maybe try to make it behave nicer in future versions.

Wow! I never thought that was the issue. Thanks for helping!

**Comment by kmamal** [comment]
I have hopefully fixed this in `v0.3.6` which I just released. The window shouldn't crash anymore no matter what image size you pass to render.
